rogue/src/player.rs

```diff
@@ -11,6 +11,226 @@ use std::cmp::{max, min};
+    fn details(&self) -> &'static str;
+}
+
+struct TryMovePlayerCommand {
+    dx: i32,
+    dy: i32,
+    codes: &'static [VirtualKeyCode],
+    description: &'static str,
+    details: &'static str,
+}
+
+
+impl KeyboundCommand for TryMovePlayerCommand {
+    fn exec(&self, ecs: &mut World) -> RunState {
+        try_move_player(self.dx, self.dy, ecs)
+    }
+
+    fn key_codes(&self) -> &'static [VirtualKeyCode] {
+        &self.codes
+    }
+
+    fn description(&self) -> &'static str {
+        self.description
+    }
+
+    fn details(&self) -> &'static str {
+        self.details
+    }
+}
+
+struct SkipTurnCommand {}
+impl KeyboundCommand for SkipTurnCommand {
+    fn exec(&self, ecs: &mut World) -> RunState {
+        skip_turn(ecs)
+    }
+
+    fn key_codes(&self) -> &'static [VirtualKeyCode] {
+        &[VirtualKeyCode::Z]
+    }
+
+    fn description(&self) -> &'static str {
+        DESCRIPTION_SKIP
+    }
+
+    fn details(&self) -> &'static str {
+        DETAILS_SKIP
+    }
+}
+
+struct PickupItemCommand {}
+impl KeyboundCommand for PickupItemCommand {
+    fn exec(&self, ecs: &mut World) -> RunState {
+        pickup_item(ecs)
+    }
+
+    fn key_codes(&self) -> &'static [VirtualKeyCode] {
+        &[VirtualKeyCode::G]
+    }
+
+    fn description(&self) -> &'static str {
+        DESCRIPTION_PICKUP
+    }
+
+    fn details(&self) -> &'static str {
+        DETAILS_PICKUP
+    }
+}
+
+
+struct DescendCommand {}
+impl KeyboundCommand for DescendCommand {
+    fn exec(&self, ecs: &mut World) -> RunState {
+        if try_next_level(ecs) {
+            return RunState::NextLevel
+        }
+        return RunState::PlayerTurn
+    }
+
+    fn key_codes(&self) -> &'static [VirtualKeyCode] {
+        &[VirtualKeyCode::Period]
+    }
+
+    fn description(&self) -> &'static str {
+        DESCRIPTION_DESCEND
+    }
+
+    fn details(&self) -> &'static str {
+        DETAILS_DESCEND
+    }
+}
+
+
+struct GenericRunstateCommand {
+    run_state: RunState,
+    codes: &'static [VirtualKeyCode],
+    description: &'static str,
+    details: &'static str,
+}
+impl KeyboundCommand for GenericRunstateCommand {
+    fn exec(&self, _ecs: &mut World) -> RunState {
+        self.run_state
+    }
+
+    fn key_codes(&self) -> &'static [VirtualKeyCode] {
+        return &self.codes
+    }
+
+    fn description(&self) -> &'static str {
+        self.description
+    }
+
+    fn details(&self) -> &'static str {
+        self.details
+    }
+}
+
+struct NoopCommand {
+    description: &'static str,
+}
+impl KeyboundCommand for NoopCommand {
+    fn exec(&self, ecs: &mut World) -> RunState {
+        RunState::AwaitingInput
+    }
+
+    fn key_codes(&self) -> &'static [VirtualKeyCode] {
+        &[]
+    }
+
+    fn description(&self) -> &'static str {
+        self.description
+    }
+
+    fn details(&self) -> &'static str {
+        self.description
+    }
+}
+
+pub const COMMANDS: [&'static dyn KeyboundCommand; 21] = [
```

Comment:
This constant is intended to capture both the association between keybindings, descriptions, and details in a structured form suitable for rendering the help menu.

In its current form, there is some measure of duplication, since the keybindings are now declared in two places (here and embedded within the pattern match statement within the `player_input(..)` function. Alternatively, that pattern match statement could be replaced by an implementation that searches this const collection to find a matching keybinding, and dispatch against the `exec(ecs) -> RunState` method on the corresponding object.

I'm planning on re-thinking this approach a little over the next few days. I like the completeness, duplicate branch warnings, and simplicity of the pattern matching solution. But the help menu benefits from some additional structure around the related help menu copy. I haven't thought of a great solution for keeping the two in sync yet.

Reply:
I assume you added the `exec` method forward thinking towards this more dynamic key dispatch? It seems like a good approach, but I do agree that not having the command => consequences written out linearly on the page is a downside.

I don't think it's much trouble at the moment to keep things in sync. If we make a mistake, maybe then it's time to make it more dynamic and DRY.